### PR TITLE
FEATURE: display commit hash for each plugin on `/admin/plugins` page.

### PR DESCRIPTION
**Pull request by vinothkannans**
It will help to find out the current version of the plugins even without the `docker_manager` plugin.

### CODE REVIEW

File: app/serializers/admin_plugin_serializer.rb [hunk 1]
```diff
@@ -10,7 +10,9 @@ class AdminPluginSerializer < ApplicationSerializer
              :enabled,
              :enabled_setting,
              :has_settings,
-             :is_official
+             :is_official,
+             :long_commit_hash,
+             :short_commit_hash
```
Comment (tgxworld) [inline app/serializers/admin_plugin_serializer.rb:15]: Is there a need to ship down both long and short commit hash from the server when the short commit hash can easily be derived from the long commit hash on the client side?

File: lib/plugin/instance.rb [hunk 2]
```diff
@@ -516,6 +516,23 @@ def after_initialize(&block)
...
   end
 
+  def commit_hash(short: false)
+    command = "git rev-parse HEAD".split(" ")
+    command = command.insert(2, "--short") if short
+    Discourse::Utils.execute_command(*command, chdir: directory).strip
+  rescue Discourse::Utils::CommandError => e
+    Rails.logger.warn("Unable to get commit hash for plugin #{name}: #{e}")
+    ""
+  end
```
Comment (tgxworld) [inline lib/plugin/instance.rb:526]: I think we can improve this further by eliminating the call to this at runtime and for each unicorn worker process by preloading this. See https://github.com/discourse/discourse/blob/e2e17bfb0f56092ccb8167b5fc39983538959d30/lib/discourse.rb#L1079

File: app/assets/javascripts/admin/addon/components/plugin-commit-hash.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import Component from "@ember/component";
+import discourseComputed from "discourse-common/utils/decorators";
+
+export default class PluginCommitHash extends Component {
```
Comment (tgxworld) [inline app/assets/javascripts/admin/addon/components/plugin-commit-hash.js:4]: I think we should be able to use glimmer components without any problem here.

**Comment by tgxworld** [comment]
@vinothkannans I tried out this locally and the commit hash doesn't seem to appear correctly on the frontend.

Screenshot from 2023-06-20 06-55-56

Actually, `commit_hash` is `nil`.

File: lib/git_repo.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    run "rev-parse HEAD"
+  end
+
+  protected
+
+  def run(cmd)
+    @memoize[cmd] ||= begin
+      return unless valid?
+      stdout, stderr, status = Open3.capture3("git #{cmd}", chdir: path)
+      status == 0 ? stdout.strip : nil
```
Comment (tgxworld) [inline lib/git_repo.rb:35]: I believe you can use [Discourse::Utils.execute_command](https://github.com/discourse/discourse/blob/c7a9da1f10848518046f696c17f16047e1cb276e/lib/discourse.rb#L24) which has certain protections like timeout built in.

File: lib/git_repo.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    end
+  rescue => e
+    puts e.inspect
```
Comment (tgxworld) [inline lib/git_repo.rb:38]: Please use `Discourse.warn_exception` here which will properly log this and capture the backtrace.

File: lib/discourse.rb [hunk 6]
```diff
@@ -1141,6 +1141,7 @@ def self.preload_rails!
         Discourse.git_version
         Discourse.git_branch
         Discourse.full_version
+        Discourse.plugins.each { |p| p.commit_url }
```
Comment (tgxworld) [inline lib/discourse.rb:1144]: What about `commit_hash`? That is also making a system call as well.

**Comment by tgxworld** [comment]
I also feel like the tests written for the changes here is insufficient. We're only asserting the existence of keys without asserting for the value. This means most of the changes that is being introduced in this PR is actually not being tested.

**Comment by vinothkannans** [comment]
> Actually, `commit_hash` is `nil`.

@tgxworld because both chat and styleguide plugins don't have separate repo. It works for other plugins. Do you think we should display the commit hash for those plugins too?

> I believe you can use [Discourse::Utils.execute_command](https://github.com/discourse/discourse/blob/c7a9da1f10848518046f696c17f16047e1cb276e/lib/discourse.rb#L24)

I just copied the [`GitRepo`](https://github.com/discourse/docker_manager/blob/eecf0af9a97d088d71a762eaeaaaf20c3a500885/lib/docker_manager/git_repo.rb#L172-L175) class from `docker_manager` plugin which we're using for a long time. Let me check whether the timeout is necessary.

> What about commit_hash?
It is already loaded to the memory when the `commit_url` is called.

**Comment by tgxworld** [comment]
> Do you think we should display the commit hash for those plugins too?

I don't know and would be better to check with product regarding this. However, the current client side code does not handle the case when `commit_hash` is `nil`.

File: lib/plugin/instance.rb [hunk 7]
```diff
@@ -516,6 +516,18 @@ def after_initialize(&block)
     initializers << block
   end
 
+  def commit_hash
+    @commit_hash ||= git_repo.latest_local_commit
+  end
+
+  def commit_url
+    @commit_url ||= "#{git_repo.url}/commit/#{git_repo.latest_local_commit}"
```
Comment (tgxworld) [inline lib/plugin/instance.rb:524]: I see that we're already memoizing stuff stuff in the `GitRepo` instance. Is there a need to memoize here again?

File: lib/plugin/instance.rb [hunk 8]
```diff
@@ -516,6 +516,18 @@ def after_initialize(&block)
     initializers << block
   end
 
+  def commit_hash
+    @commit_hash ||= git_repo.latest_local_commit
```
Comment (tgxworld) [inline lib/plugin/instance.rb:520]: Same as https://github.com/discourse/discourse/pull/22176/files#r1235997289

File: lib/discourse.rb (hunk 6 above)
Comment (vinothkannans) [reply lib/discourse.rb:1144]: Now, the `commit_url` method will call the `commit_hash`

**Comment by vinothkannans** [comment]
@tgxworld now all the changes requested are completed.

File: app/assets/javascripts/admin/addon/components/plugin-commit-hash.hbs [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<a
+  href={{this.plugin.commit_url}}
+  class="current commit-hash"
+  title={{this.plugin.commit_hash}}
+>{{this.shortCommitHash}}</a>
```
Comment (tgxworld) [inline app/assets/javascripts/admin/addon/components/plugin-commit-hash.hbs:5]: I still think my concern where `commit_hash` is `nil` needs to be addressed. When `commit_hash` is `nil`, we're generating a some what invalid link.

Comment (vinothkannans) [reply app/assets/javascripts/admin/addon/components/plugin-commit-hash.hbs]: I added an "if" condition. Now it won't display the invalid link if `commit_hash` is not available.

**Review by tgxworld — APPROVED** [review]

**Comment by tgxworld** [comment]
Looks good :+1: Thank you for the changes.

**Pull request merged**